### PR TITLE
DOC-9281: Update Error Handling page for Python SDK

### PR DESCRIPTION
**Pull request by thejcfactor**

### CODE REVIEW

File: modules/howtos/examples/error_handling.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
+
```
Comment (osfameron) [inline modules/howtos/examples/error_handling.py:1]: it looks like there were significant changes to this file, and not only removing the swathe of .adoc included in it, so I think it makes sense to review as if this was an entirely new code example?

File: modules/howtos/examples/error_handling.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        @functools.wraps(func)
+        def func_wrapper(*args, **kwargs):
+            for i in range(retry_limit):
+                try:
+                    return func(*args, **kwargs)
+                except Exception as ex:
+                    if allowed_exceptions is None or not isinstance(ex, allowed_exceptions):
+                        raise
+
+                    if (retry_limit-1)-i == 0:
```
Comment (osfameron) [inline modules/howtos/examples/error_handling.py]: NITPICK: this feels a bit clunky.
`range(1, retry_limit+1)` above would remove the awkward subtraction here?
Also subtract and compare 0 is a bit odd.

File: modules/howtos/examples/error_handling.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        print(ex.context.statement)
+        print(ex.context.first_error_code)
+        print(ex.context.first_error_message)
+        print(ex.context.client_context_id)
+        print(ex.context.endpoint)
+
+# end::QueryErrorContext[]
+
+try:
+    collection.remove("my-key")
```
Comment (osfameron) [inline modules/howtos/examples/error_handling.py:152]: comment this as teardown perhaps?

File: modules/howtos/pages/error-handling.adoc [hunk 4]
```diff
@@ -63,141 +71,85 @@ Retry with a linearly increasing delay.
 Retry with an exponentially increasing delay.
 Retry with a random delay.
 
+Consider decorator that provides flexibility to determine which Exceptions to retry and how to retry (fixed, linear or exponential delay).
```
Comment (osfameron) [inline modules/howtos/pages/error-handling.adoc]: "Consider *a* decorator"

File: modules/howtos/pages/error-handling.adoc [hunk 5]
```diff
@@ -63,141 +71,85 @@ Retry with a linearly increasing delay.
...
-=== Durability ambiguous
+=== Ambiguity
 
 There are situations with any distributed system in which it is simply impossible to know for sure if the operation completed successfully or not.
 Take this as an example: your application requests that a new document be created on Couchbase Server.
 This completes, but, just before the server can notify the client that it was successful, a network switch dies and the application's connection to the server is lost.
 The client will timeout waiting for a response and will raise a `TimeoutException`, but it's ambiguous to the app whether the operation succeeded or not.
 
-So `ErrTimeout` is one ambiguous error, another is `ErrDurabilityAmbiguous`, which can returned when performing a durable operation.
+So `TimeoutException` is one ambiguous exception, another is `DurabilitySyncWriteAmbiguousException`, which can returned when performing a durable operation.
```
Comment (osfameron) [inline modules/howtos/pages/error-handling.adoc]: split sentence.

File: modules/howtos/pages/error-handling.adoc [hunk 6]
```diff
@@ -63,141 +71,85 @@ Retry with a linearly increasing delay.
...
 ----
 
-=== Durability invalid level
+=== Non-Idempotent Operations
 
-[source,python]
-----
-include::example$error-handling.py[tag=DurabilityInvalidLevelException]
-----
+Idempotent operations are those that can be applied multiple times yet still have the one, same effect. Repeatedly setting an email field is idempotent increasing a counter by one is not.
```
Comment (osfameron) [inline modules/howtos/pages/error-handling.adoc]: comma (or new sentence before "increasing a counter"

File: modules/howtos/examples/error_handling.py (hunk 3 above)
Comment (osfameron) [reply modules/howtos/examples/error_handling.py:152]: done

File: modules/howtos/pages/error-handling.adoc (hunk 4 above)
Comment (osfameron) [reply modules/howtos/pages/error-handling.adoc]: done

File: modules/howtos/pages/error-handling.adoc (hunk 6 above)
Comment (osfameron) [reply modules/howtos/pages/error-handling.adoc]: Tweaked this section.

**Review by osfameron — APPROVED** [review]
Looks great thanks!

This resolves the original question in DOC-9281. I'll ticket some more work on error handling in general for future.

I left one minor NITPICK on code, let me know what you think, and feel free to update if you agree,
otherwise I'll merge shortly in any case.

**Comment by thejcfactor** [comment]
Gosh, I need to figure out what is up with my alerts.  Thanks @osfameron! I will check this out tomorrow and provide updates/feedback.

File: modules/howtos/examples/error_handling.py (hunk 1 above)
Comment (thejcfactor) [reply modules/howtos/examples/error_handling.py:1]: I would agree.  TBH, I try to swap out older examples for new/improved code examples whenever I am working on the docs.  Is there something we need to do differently when it is a new code example?

File: modules/howtos/examples/error_handling.py (hunk 2 above)
Comment (thejcfactor) [reply modules/howtos/examples/error_handling.py]: Let me know what you think about the change.  Good catch -- that was not clean code :(

File: modules/howtos/examples/error_handling.py (hunk 3 above)
Comment (thejcfactor) [reply modules/howtos/examples/error_handling.py:152]: LGTM

File: modules/howtos/pages/error-handling.adoc (hunk 4 above)
Comment (thejcfactor) [reply modules/howtos/pages/error-handling.adoc]: LGTM

File: modules/howtos/pages/error-handling.adoc (hunk 5 above)
Comment (thejcfactor) [reply modules/howtos/pages/error-handling.adoc]: LGTM

File: modules/howtos/pages/error-handling.adoc (hunk 6 above)
Comment (thejcfactor) [reply modules/howtos/pages/error-handling.adoc]: LGTM

File: modules/howtos/examples/error_handling.py (hunk 1 above)
Comment (osfameron) [reply modules/howtos/examples/error_handling.py:1]: No, it's fine! Just wanted to clarify my approach as I couldn't figure out what had changed with a diff.

Doing rename as a separate commit (or brief mention in commit message) might have saved me a few minutes I guess? Not a biggie.

**Review by osfameron — APPROVED** [review]
looks great 👍

**Pull request merged**